In [7]:
# import required packages
import numpy as np
# load fft package
from numpy import fft

# Problem 4

In [8]:
# define a variable f
# set it to be an periodic array
f = np.array([0.3679, 1.0000, 2.7183, 4.1133, 2.7183, 1.0000, 0.3679, 0.2431])

# define a variable x
# set it to be the length of the periodic array f
N = len(f)

# use fft to compute the Fourier coefficients 
fc = fft.fft(f)

# compute a_k for all k=0,1,...,N/2
a = 1/N*(fc + fc.conj())
a = np.real(a)[0:N//2+1]
print('a_k =', a)

# compute b_k for all k=0,1,...,N/2-1
b = 1j/N*(fc - fc.conj())
b = np.real(b)[0:N//2]
print('b_k =', b)

a_k = [ 3.1322     -1.27176117  0.          0.09656117 -0.046     ]
b_k = [ 0.          1.27176117 -0.5891      0.09656117]


Hence,
$$
\begin{aligned}
P_{8}(x)=\\
\approx&1.5661-1.2718\cos(x)+1.2718\sin(x)-0.5891\sin(2x)\\
&+0.0966 \cos(3x)+0.0966 \sin(3x)-0.046\cos(4x)
\end{aligned}
$$


# Problem 5

In [10]:
N = 8

# define variables x and f
x = np.arange(N)*2*np.pi/N # use np.arange to return even-spaced values within the indicated interval
f = np.exp(np.cos(x))# exp will calulate the exponential of items in the array x

# # use fft to compute the Fourier coefficients 
fc = fft.fft(f)

# define variables a and b 
a = np.real(1/N*(fc + fc.conj()))[0:N//2+1] # use np.real to return the real part
b = np.real(1j/N*(fc - fc.conj()))[0:N//2] # use np.real to return the real part

# compute P'8(xj)
dP = np.zeros_like(f) # use np.zeros_like to return an array of zeros that has the same type as f
for j, xj in enumerate(x): # implement a for loop with indicator variables j and xj
    k = np.arange(1,N//2) # use np.arange to return even-spaced values within the indicated interval
    dP[j] = sum(k*(-a[k]*np.sin(k*xj)+b[k]*np.cos(k*xj))) - 2*a[N//2]*np.sin(N/2*xj)
    
# computation for the actual error
error = -np.exp(np.cos(x))*np.sin(x) - dP
print("The actual error in the approximation f'(xj) - P'(xj) is\n", error)

The actual error in the approximation f'(xj) - P'(xj) is
 [-4.25040794e-16  3.44909971e-03 -4.31791110e-03  2.72947135e-03
 -1.99533261e-17 -2.72947135e-03  4.31791110e-03 -3.44909971e-03]
